### PR TITLE
feat: decode normal packets

### PR DESCRIPTION
**Pull request by ChillerDragon**
BTW if you are curious what kind of bytes are used in the test case

```C++
	uint8_t bytes[] = {
	0x00, 0x06, 0x02, 0x42, 0x0d, 0x05, 0x28, 0x43,
	0x68, 0x69, 0x6c, 0x6c, 0x65, 0x72, 0x44, 0x72,
	0x61, 0x67, 0x6f, 0x6e, 0x00, 0x7c, 0x2a, 0x4b,
	0x6f, 0x47, 0x2a, 0x7c, 0x00, 0x80, 0x01, 0x67,
	0x72, 0x65, 0x65, 0x6e, 0x73, 0x77, 0x61, 0x72,
	0x64, 0x00, 0x00, 0x87, 0xc5, 0x8d, 0x0e, 0x8e,
	0xab, 0x9e, 0x02, 0x40, 0x0c, 0x06, 0x23, 0x63,
	0x72, 0x61, 0x73, 0x68, 0x6d, 0x65, 0x70, 0x6c,
	0x78, 0x00, 0x3d, 0xe3, 0x94, 0x8d
	};
```

You can copy paste them as is into the [web traffic decoder](https://twnet.zillyhuhn.com/?d=00+06+02+42+0d+05+28+43+68+69+6c+6c+65+72+44+72+61+67+6f+6e+00+7c+2a+4b+6f+47+2a+7c+00+80+01+67+72+65+65+6e+73+77+61+72+64+00+00+87+c5+8d+0e+8e+ab+9e+02+40+0c+06+23+63+72+61+73+68+6d+65+70+6c+78+00+3d+e3+94+8d&v=6) to see the parsed  packet as json.

image

image

### CODE REVIEW

**Comment by MilkeeyCat** [comment]
I think it would be easier and better to write a script which decodes game and system messages. I started working on it earlier but didn't finish.

**Comment by ChillerDragon** [comment]
I disagree that it would be easier. For me writing a script has always been slower than doing it by hand. Also the script is more complicated to maintain if you want to make changes to the generated code. The main advantage of a script is that it can in the future easily generate new messages that were added by ddnet.

But script or not let's write out 1 or two messages by hand and merge it to master just to see if the structure works.
If it is good we can then generate that exact code. If it is bad we can change it before we start the generation.

**Comment by ChillerDragon** [comment]
Could you open a draft pr with your script? I am curious how it looks like.

**Comment by ChillerDragon** [comment]
Went for the ``msg.rcon_cmd`` instead of ``msg.msg_rcon`` https://github.com/MilkeeyCat/ddnet_protocol/issues/28#issuecomment-2525253163

I think naming the union is nicer. Because a chunk has a header and a message. So I expect these two fields in the struct.

**Comment by MilkeeyCat** [comment]
Actually I think we can raw dawg those messages without generating them.

**Comment by ChillerDragon** [comment]
> Actually I think we can raw dawg those messages without generating them.

By raw dawg do you mean not provide structs at all and let the users do

```C
Packer packer;
packer_init_msg(&packer, MSG_RCON_CMD, MESSAGE_SYSTEM);
packer_add_string(&packer, "say hello");
```

Or use structs but not use code generation at all and write all of them by hand? If that is the case is there anything still blocking this pr?

File: src/packet_normal.c [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+#include "message.h"
+#include "packet.h"
+
+PacketNormal *decode_normal(uint8_t *buf, size_t len, PacketHeader header, Error *err) {
+	PacketNormal *packet = malloc(sizeof(PacketNormal));
+	packet->_ = PACKET_NORMAL;
+	packet->header = header;
+
+	uint8_t *end = buf + len;
+	int32_t num_chunks = 0;
```
Comment (MilkeeyCat) [inline src/packet_normal.c:12]: Use `size_t`

**Review by MilkeeyCat — CHANGES_REQUESTED** [review]

File: src/packet_normal.c (hunk 1 above)
Comment (ChillerDragon) [reply src/packet_normal.c]: Oh yea true. I do not have the habit of using anything but int :D
Compiler warnings and casting have trained me to go int everywhere and live an easy life. But it indeed seems more correct conceptually.

I noticed the packet header uses ``uint8_t`` so I went with that for consistency.

https://github.com/MilkeeyCat/ddnet_protocol/blob/3a1d4c4403b9050d8deb8ce5faffcc87347b0a40/include/ddnet_protocol/packet.h#L38

**Comment by MilkeeyCat** [comment]
Thanks!

**Pull request merged**